automation/orchestrator_test.go

```diff
@@ -61,74 +48,179 @@ func TestIntegration(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer o.Shutdown()
 
-	rtt1 := trigger.NewRuntimeTrigger()
-	rtt2 := trigger.NewRuntimeTrigger()
-	rtt2.SetActive(true)
+	prevTriggerNewID := trigger.NewID
+	defer func() {
+		trigger.NewID = prevTriggerNewID
+	}()
+	triggerIDIndex := 0
+	triggerIDs := [2]string{
+		"id1",
+		"id2",
+	}
+	trigger.NewID = func() string {
+		if triggerIDIndex >= len(triggerIDs) {
+			t.Fatal("trigger.NewID called more times then expected")
+		}
+		id := triggerIDs[triggerIDIndex]
+		triggerIDIndex++
+		return id
+	}
 	expected := &workflow.Workflow{
 		DatasetID: "dataset_id",
 		OwnerID:   "profile_id",
 		Created:   NowFunc(),
-		Triggers:  []trigger.Trigger{rtt1, rtt2},
+		Triggers: []map[string]interface{}{
+			map[string]interface{}{
+				"id":           triggerIDs[0],
+				"type":         trigger.RuntimeType,
+				"active":       false,
+				"advanceCount": 0,
+			},
+			map[string]interface{}{
+				"id":           triggerIDs[1],
+				"type":         trigger.RuntimeType,
+				"active":       true,
+				"advanceCount": 0,
+			}},
 	}
 
-	triggerOpts := []map[string]interface{}{
-		map[string]interface{}{"type": trigger.RuntimeType},
-		map[string]interface{}{"type": trigger.RuntimeType, "active": true},
-	}
-
-	allowUnexported := cmp.AllowUnexported(trigger.RuntimeTrigger{})
-	got, err := o.CreateWorkflow("dataset_id", "profile_id", triggerOpts)
+	got, err := o.SaveWorkflow(&workflow.Workflow{
+		DatasetID: "dataset_id",
+		OwnerID:   "profile_id",
+		Triggers: []map[string]interface{}{
+			map[string]interface{}{"type": trigger.RuntimeType},
+			map[string]interface{}{"type": trigger.RuntimeType, "active": true},
+		}})
 	if err != nil {
 		t.Fatal(err)
 	}
 	expected.ID = got.ID
-	if diff := cmp.Diff(expected, got, allowUnexported); diff != "" {
+	if diff := cmp.Diff(expected, got); diff != "" {
 		t.Errorf("workflow mismatch (-want +got):\n%s", diff)
 	}
 
-	if runtimeListener.TriggerExists(expected) {
+	if runtimeListener.TriggersExists(expected) {
 		t.Fatal("only triggers of active workflows should be added to the runtimeListener")
 	}
 
 	got, err = o.GetWorkflow(expected.ID)
 	if err != nil {
 		t.Fatal(err)
 	}
-	if diff := cmp.Diff(expected, got, allowUnexported); diff != "" {
+	if diff := cmp.Diff(expected, got); diff != "" {
 		t.Errorf("workflow mismatch (-want +got):\n%s", diff)
 	}
 
+	runID := "runID_1"
+	expectedWorkflowStartedEvent := event.WorkflowStartedEvent{
+		DatasetID:  got.DatasetID,
+		OwnerID:    got.OwnerID,
+		WorkflowID: got.WorkflowID(),
+		RunID:      runID,
+	}
+	expectedWorkflowStoppedEvent := event.WorkflowStoppedEvent{
+		DatasetID:  got.DatasetID,
+		OwnerID:    got.OwnerID,
+		WorkflowID: got.WorkflowID(),
+		RunID:      runID,
+		Status:     string(run.RSSucceeded),
+	}
+	var gotWorkflowStartedEvent event.WorkflowStartedEvent
+	var gotWorkflowStoppedEvent event.WorkflowStoppedEvent
+	workflowStoppedEventFired := make(chan string)
```

Comment:
realized we have 2 separate async events we need to make sure are happening in order to test properly
1) did the runFunc actually get triggered
2) did the workflow events fire properly. 

We need to wait for verification that the workflow events finished firing before we can test that they fired with all the correct information

Reply:
agreed. I think the better way to enforce this is to list out an expected order of events in a slice, build up a slice of all events received as you receive them, and compare the two

Reply:
ah, yes I'll definitely refactor to collect in a slice, but I phrased the above wrong. By "two separate events" I meant "two separate actions" 1) that the runFunc gets called and 2) that we have the correct events. But now looking at it, checking for the run func was only necessary before we put the events in place. Going to refactor this whole section.